Add integration tests

ZachEddy opened the pull request:
# Add Integration Tests

This pull request adds integration tests to the contacts application. For the time being, not all endpoints have been tested, but I was eager to push some code. 

The challenging part was figuring out how to launch the contacts application and reset the Postgres database between tests. At this point, I have it pretty well figured out, so it should hopefully be straightforward to add new tests.

Here's a brief summary of the major changes I've made.

- **Create the `MigrateDB` function** — I needed a way to run migrations repeatedly, so I moved the migration code to its own exported function. Also worth mentioning: the `MigrateDB` function takes a `sql.DB` type, so it can be reworked to use something other than `gorm.AutoMigrate` without breaking the tests.

- **Running a Test Postgres Container** — Contributors to the contacts application don't have to configure a local Postgres database to run integration tests. When the integration tests run, a Postgres container will automatically spin-up. The gRPC server will connect to this Postgres container, some tests will run, then the container will exit and delete itself.

- **Building and Running Binaries** — The integration `TestMain` will build the gateway and server binaries, run them as separate processes (for the time being), then delete the binaries when finished.

Calebjh commented on .travis.yml:
```diff
@@ -0,0 +1,25 @@
+sudo: required
```
We bounced this as an idea for fixing an error before, is it still required?

ZachEddy replied:
Yeah, it's required: see this [TravisCI](https://docs.travis-ci.com/user/docker/) documentation.

dmacthedestroyer commented on integration/database.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+// PostgresDBConfig models a test Postgres database
+type PostgresDBConfig struct {
```
or specify port `0` and the OS will choose one for you:
https://gist.github.com/xcsrz/538e291d12be6ee9a8c7
https://gist.github.com/xcsrz/538e291d12be6ee9a8c7

dmacthedestroyer commented on integration/main_integration_test.go:
```diff
@@ -0,0 +1,86 @@
+			packagePath, err, string(out),
+		)
+	}
+	return func() {
```
the difference in my mind is: is this going to be library code, or application code?

Library code should always return the error instead of `log.Fatal`. This gives the consumer the option of how they want to deal with the error.

This could be something we discuss further when we get into what can be generalized into the toolkit.

dmacthedestroyer commented on integration/database.go:
```diff
@@ -0,0 +1,75 @@
+
+// RunAsContainer uses the Postgres configuration to run a Postgres Docker
+// container locally
+func (cfg PostgresDBConfig) RunAsContainer(containerName string) func() {
```
hmm, I think it's fine to have it take in a `testing.T` because that implies that it should be called on a per-test basis, which it should.

If you wanted to be fancy, you could have two versions of the method -- one that returns an `error` and one that takes in a `testing.T`. I'm not opinionated about that and don't have the interest to come up with names for each :)

ZachEddy commented:
Alright, I updated the pull request with @dmacthedestroyer's feedback. I did a bad job of adding commits as a went, so here's an overview of the changes.

- Remove `log.Fatal` calls inside the "library" code. Errors are now handled in `MainTest` instead.
- Moved the building-and-running code into a file called `exec.go` to help organize the package.
- Wrote a generic `RunContainer` function inside `exec.go` that runs a user-specified image. I did this to keep the `exec.Cmd` code contained in one spot. The `PostgresDBConfig` calls this function to spin-up a new container.
- Added the `transport.go` file (please suggest a better name if you can think of one). It includes the `GetOpenPortInRange` function, which helps avoid hard-coding a port number.

Calebjh commented on integration/grpc_test.go:
```diff
@@ -0,0 +1,122 @@
+// contacts application
+// 1. Create a contact
+// 2. Read contact with the returned ID from the response to ensure it exists
+// 3. Ensure the contact was correctly deleted
```
I think a step is missing in this description
